sos/plugins/__init__.py

```diff
@@ -1026,6 +1026,15 @@ def report(self):
         else:
             return html
 
+    def check_process_by_name(self, process):
+        """Checks if a named process is found in full command line using pgrep.
+        Returns either True or False."""
+        ps = self.get_command_output("pgrep -f %s" % process, timeout=0)
```

Comment:
Shall not we use `"pgrep -fx %s"` there?

       -x, --exact
              Only match processes whose names (or command line if -f is specified) exactly match the pattern.

Just to ensure we dont get false positives like:

    # ps aux | grep qpidd
    qpidd     1068  0.1  0.2 495532 33012 ?        Ssl  Jan09   2:57 /usr/sbin/qpidd --config /etc/qpid/qpidd.conf
    root     12397  0.0  0.0 113128  1184 pts/1    S    08:30   0:00 bash monitor_qpidd.sh
    root     12402  0.0  0.0 112664   980 pts/1    S+   08:30   0:00 grep --color=auto qpidd
    # pgrep -f qpidd
    1068
    12397
    # kill 1068
    # ps aux | grep qpidd
    root     12397  0.0  0.0 113128  1184 pts/1    S    08:30   0:00 bash monitor_qpidd.sh
    root     12419  0.0  0.0 112660   980 pts/1    S+   08:30   0:00 grep --color=auto qpidd
    # pgrep -f qpidd
    12397
    # echo $?
    0
    # 

Here no "`qpidd`" process is running, but `check_process_by_name` returns `True`.

Reply:
with -x you'd need to specify the full process command. like:

~~~
# ps aux  |grep cinder_wsgi
cinder    552235  0.2  1.1 1142484 137320 ?      Sl   Jan10   1:17 cinder_wsgi     -DFOREGROUND
cinder    552236  0.2  1.1 1142484 137268 ?      Sl   Jan10   1:15 cinder_wsgi     -DFOREGROUND
root      629453  0.0  0.0 112660   972 pts/4    S+   07:52   0:00 grep --color=auto cinder_wsgi

# pgrep -fx cinder_wsgi
# echo $?
1

# pgrep -fx "cinder_wsgi     -DFOREGROUND"
552235
552236
# echo $?
0
~~~

But I agree with just -f we might have false positives. On the other side plugins might need modifications more often if command options get modified from release to release, or if custom options can get set for the service.

Reply:
Ah I was wrong; `pgrep -fx qpidd` really does *not* find the process with cmdline `/usr/sbin/qpidd --config /etc/qpid/qpidd.conf`. So `-x` makes little sense.

Reply:
Or we could just do it ourselves in Python... It's almost as simple, and it is often faster and less resource hungry:

```python
import sys
from glob import glob
from time import time

comm_glob = "/proc/[0-9]*/comm"

def main(argv):
    start_time = time()
    comm_match = argv[1]
    comm_paths = glob(comm_glob)
    for path in comm_paths:
        with open(path, "r") as f:
            comm = f.read().strip()
            if comm_match == comm:
                print("'%s' is running" % comm_match)
                print("t=%f" % (time() - start_time))
                sys.exit(0)
    print("'%s' not running" % comm_match)
    print("t=%f" % (time() - start_time))
    sys.exit(1)

if __name__ == "__main__":
    main(sys.argv)
```

```bash
$ python /tmp/proctest.py vim
'vim' is running
t=0.005336
$ python /tmp/proctest.py qux
'qux' not running
t=0.009143
$ time pgrep qux
real	0m0.025s
user	0m0.005s
sys	0m0.019s
$ time pgrep vim
7257
7757
18268

real	0m0.014s
user	0m0.002s
sys	0m0.012s
```

Reply:
Similar benefit with > 10,000 processes running:

```bash
$ for i in {0..10000}; do sleep 240; done
[...]
$ time pgrep vim
7257
7757
18268

real	0m0.218s
user	0m0.036s
sys	0m0.180s
$ python /tmp/proctest.py qux
'qux' not running
t=0.156281
$ python /tmp/proctest.py vim
'vim' is running
t=0.095509
$ python /tmp/proctest.py qux
'qux' not running
t=0.144969
$ python /tmp/proctest.py vim
'vim' is running
t=0.087021
$ python /tmp/proctest.py qux
'qux' not running
t=0.144287
$ python /tmp/proctest.py vim
'vim' is running
t=0.086775
$ time pgrep vim
7257
7757
18268

real	0m0.211s
user	0m0.033s
sys	0m0.177s
```

Reply:
I have changed the method, but check the `/proc/[0-9]*/cmdline` as e.g. with services running
via httpd wsgi we'd only see httpd in `/proc/[0-9]*/cmdline` instead of the process name we are
interested in. If in that case httpd is for interest, we'd still see the master httpd process.

Reply:
> instead of the process name we are interested in. If in that case httpd is for interest, we'd still see the 
> master httpd process.

 Are you confusing the process _name_ (`/proc/<pid>/comm`) with some of its arguments (`/proc/<pid>/cmdline')?

The value stored in `comm` is the process name. If that's not what you need to key on then you need a different method name at least - `check_process_by_name()` is a highly misleading name if you are actually checking anything other than the value of `comm`.

Reply:
In case of apache mod_wsgi what is shown by `ps` can be changed using the `display-name`

~~~
[root@controller-0 ~]# grep display-name /etc/httpd/conf.d/10-cinder_wsgi.conf 
  WSGIDaemonProcess cinder-api display-name=cinder_wsgi group=cinder processes=4 threads=4 user=cinder
~~~

From - http://modwsgi.readthedocs.io/en/develop/configuration-directives/WSGIDaemonProcess.html:

~~~
display-name=value
Defines a different name to show for the daemon process when using the ps command to list processes. If the value is %{GROUP} then the name will be (wsgi:group) where group is replaced with the name of the daemon process group.
~~~

With this we see cinder_wsgi in the ps process output, what the user tries to find as name from his plugin:

~~~
[root@controller-0 ~]# ps -ef |grep cinder_wsgi
cinder    157920  157916  0 12:36 ?        00:00:00 cinder_wsgi     -DFOREGROUND
cinder    157921  157916  0 12:36 ?        00:00:00 cinder_wsgi     -DFOREGROUND
cinder    157922  157916  0 12:36 ?        00:00:00 cinder_wsgi     -DFOREGROUND
cinder    157923  157916  0 12:36 ?        00:00:00 cinder_wsgi     -DFOREGROUND
root      621087  620810  0 19:55 pts/0    00:00:00 grep --color=auto cinder_wsgi
~~~

But `/proc/157920/comm`: 

~~~
[root@controller-0 ~]# cat /proc/157920/comm 
httpd
~~~

We can get the process by name from cmdline:

~~~
[root@controller-0 ~]# cat /proc/157920/cmdline 
cinder_wsgi    -DFOREGROUND
~~~

Since `/proc/pid/cmdline` holds the complete command line we can get the information someone expects from `ps` output from there. With the downside that we might have false positives as discussed with @pmoravec before.